ocs/agent/aggregator.py

```diff
@@ -149,10 +150,174 @@ def empty(self):
+                                   field=field_name)
+                    verified = False
+
+        return verified
+
+    @staticmethod
+    def _enforce_field_name_rules(field_name):
+        """Enforce naming rules for field names.
+
+        A valid name:
+
+        * contains only letters (a-z, A-Z; case sensitive), decimal digits (0-9), and the
+          underscore (_).
+        * begins with a letter, or with any number of underscores followed by a letter.
+        * is no more than 255 characters long.
+
+        Args:
+            field_name (str):
+                Field name string to check and modify if needed.
+
+        Returns:
+            str: New field name, meeting all above rules. Note this isn't
+                 guarenteed to not collide with other field names passed
+                 through this method, and that should be checked.
+
+        """
+        # replace spaces with underscores
+        new_field_name = field_name.replace(' ', '_')
+
+        # replace invalid characters
+        new_field_name = re.sub('[^a-zA-Z0-9_]', '', new_field_name)
+
+        # grab leading underscores
+        underscore_search = re.compile('^_*')
+        underscores = underscore_search.search(new_field_name).group()
+
+        # remove leading underscores
+        new_field_name = re.sub('^_*', '', new_field_name)
+
+        # remove leading non-letters
+        new_field_name = re.sub('^[^a-zA-Z]*', '', new_field_name)
+
+        # add underscores back
+        new_field_name = underscores + new_field_name
+
+        # limit to 255 characters
+        new_field_name = new_field_name[:255]
+
+        return new_field_name
+
+    @staticmethod
+    def _check_for_duplicate_names(field_name, name_list):
+        """Check name_list for matching field names and modify field_name if
+        matches are found.
+
+        The results of Provider._enforce_field_name_rules() are not guarenteed
+        to be unique. This method will check field_name against a list of field names
+        of existing field names and try to append '_N', with N being a zero
+        padded integer up to 99.
+
+        In the event the field name is at the maximum allowed length, we remove
+        some characters before appending the additional underscore and integer.
+
+        Examples:
+            >>> current_field_names = ['test', 'test_01']
+            >>> name = 'test'
+            >>> new_name = Provider._check_for_duplicate_names(name, current_field_names)
+            >>> print(new_name)
+            test_02
+
+        Args:
+            field_name (str): field name to check against name_lsit
+            name_list (list): list of field names already in a Block
+
+        Returns:
+            str: A new field name that is not already in name_list
+
+        """
+        # Assume duplicate name, until proven otherwise
+        duplicate_name = True
+        name_index = 1
+
+        while duplicate_name:
```

Comment:
Right, that does make it more readable, will change.

Good point about the index becoming large enough to push the name over length limits. Updated to build the suffix and check its length, removing that many characters from the end of the string instead of a fixed three.

Reply:
I believe the latest version is still a little off -- instead of hard-coded 252 you want 255 - suf_len.

Reply:
Right, fixed.